File: Anole/HTMLRenderBand.cs
using System;
using System.Collections;
using System.Drawing;

namespace Anole
{
	/// <summary>
	/// Summary description for HTMLRenderBand.
	/// </summary>
	public class HTMLRenderBand
	{
		protected ArrayList myRenderItems;
		protected int myTargetWidth;
		public int TargetWidth
		{
			get{return myTargetWidth;}
			set{myTargetWidth=value;}
		}

		protected int myWidth;
		public int Width
		{
			get{return myWidth;}
			set{myWidth=value;}
		}

		protected int myHeight;
		public int Height
		{
			get{return myHeight;}
			set{myHeight=value;}
		}

		protected int myTop;
		public int Top
		{
			get{return myTop;}
			set{myTop=value;}
		}

		public HTMLRenderBand(int targetwidth)
		{
			myTargetWidth=targetwidth;
			myRenderItems=new ArrayList();
			myHeight=0;
			myWidth=0;
		}

		public int WidthRemaining()
		{
			if(myWidth<myTargetWidth)
			{
				return myTargetWidth-myWidth;
			}
			return 0;
		}

		public int GetItemCount()
		{
			return myRenderItems.Count;
		}

		public void AddItem(HTMLRenderItem item)
		{
			myHeight=Math.Max(myHeight,item.Height);
			myWidth+=item.Width;
			myRenderItems.Add(item);
		}

		/// <summary>
		/// Called by HTMLRenderer after bands are filled
		/// with RenderItems so alignment can be calculated
		/// </summary>
		public void ComputeItemCoords()
		{
			int left=1;

			//TBD: valign for table cells

			if(myRenderItems.Count==0) return;

			//If an item in the band is centered, then
			//assume all items in that band are centered
			HTMLRenderItem hri=myRenderItems[0] as HTMLRenderItem;
			if(hri.Element.Alignment==HTMLAlignment.Center)
				left=(this.TargetWidth-this.Width)/2;
			else
				left=1;
	
			foreach(HTMLRenderItem item in myRenderItems)
			{
				item.RenderBand=this;
				item.Left=left;
				item.Top=this.Top;
				left+=item.Width;
			}
			
		}

		public void Paint(Graphics g)
		{
			/*
			Pen p=new Pen(Color.Black);
			g.DrawRectangle(p,0,this.Top,this.Width,this.Height);
			g.FillRectangle(new SolidBrush(Color.White),1,this.Top+1,
				this.Width-2,this.Height-2);
				*/
			foreach(HTMLRenderItem item in myRenderItems)
			{
				item.Paint(g);
			}
		}

		/// <summary>
		/// Returns the item whose
		/// bounding rectangle contains the specified point.
		/// Used for hyperlink hit-testing.
		/// </summary>
		/// <param name="x"></param>
		/// <param name="y"></param>
		/// <returns></returns>
		public HTMLRenderItem LocateItem(int x, int y)
		{
			foreach(HTMLRenderItem item in this.myRenderItems)
			{
				if((x>=item.Left)&&(x<=item.Left+item.Width))
				{
					if((y>=item.Top) && (y<=item.Top+item.Height))
					{
						return item;
					}
				}
				
			}
			return null;
		}

		public HTMLRenderItem LocateItem(string aname)
		{
			foreach(HTMLRenderItem item in myRenderItems)
			{
				if(item.Element.Name.Length>0)
				{
					if(item.Element.Name.CompareTo(aname)==0)
					{
						return item;
					}
				}
			}
			return null;
		}

		public override string ToString()
		{
			string s="";
			s="<HTMLRenderBand top:" + this.Top.ToString() + " width:" + this.Width.ToString()+" height:"+this.Height.ToString()+" ";
			foreach(HTMLRenderItem item in myRenderItems)
			{
				s+=item.ToString();
			}
			s+=" HTMLRenderband>";
			return s;
		}



	}
}

File: Anole/HTMLRenderBand.vb
Imports System
Imports System.Collections
Imports System.Drawing

Namespace Anole

    Public Class HTMLRenderBand

        Protected myRenderItems As ArrayList
        Protected myTargetWidth As Integer

        Public Property TargetWidth() As Integer
            Get
                Return myTargetWidth
            End Get
            Set(ByVal value As Integer)
                myTargetWidth = Value
            End Set
        End Property

        Protected myWidth As Integer

        Public Property Width() As Integer
            Get
                Return myWidth
            End Get
            Set(ByVal value As Integer)
                myWidth = Value
            End Set
        End Property

        Protected myHeight As Integer

        Public Property Height() As Integer
            Get
                Return myHeight
            End Get
            Set(ByVal value As Integer)
                myHeight = Value
            End Set
        End Property

        Protected myTop As Integer

        Public Property Top() As Integer
            Get
                Return myTop
            End Get
            Set(ByVal value As Integer)
                myTop = Value
            End Set
        End Property

        Public Sub New(ByVal targetwidth As Integer)
            myTargetWidth = targetwidth
            myRenderItems = New ArrayList()
            myHeight = 0
            myWidth = 0
        End Sub

        Public Function WidthRemaining() As Integer
            If myWidth < myTargetWidth Then
                Return myTargetWidth - myWidth
            End If
            Return 0
        End Function

        Public Function GetItemCount() As Integer
            Return myRenderItems.Count
        End Function

        Public Sub AddItem(ByVal item As HTMLRenderItem)
            myHeight = Math.Max(myHeight, item.Height)
            myWidth += item.Width
            myRenderItems.Add(item)
        End Sub

        ' Called by HTMLRenderer after bands are filled
        ' with RenderItems so alignment can be calculated
        Public Sub ComputeItemCoords()
            Dim left As Integer = 1

            'TBD: valign for table cells

            If myRenderItems.Count = 0 Then Return

            'If an item in the band is centered, then
            'assume all items in that band are centered
            Dim hri As HTMLRenderItem = TryCast(myRenderItems(0), HTMLRenderItem)

            If hri.Element.Alignment = HTMLAlignment.Center Then
                left = (Me.TargetWidth - Me.Width) \ 2
            Else
                left = 1
            End If

            For Each item As HTMLRenderItem In myRenderItems
                item.RenderBand = Me
                item.Left = left
                item.Top = Me.Top
                left += item.Width
            Next

        End Sub

        Public Sub Paint(ByVal g As Graphics)

            '   Pen p=new Pen(Color.Black);
            '	g.DrawRectangle(p,0,this.Top,this.Width,this.Height);
            '	g.FillRectangle(new SolidBrush(Color.White),1,this.Top+1,
            '	this.Width-2,this.Height-2);			

            For Each item As HTMLRenderItem In myRenderItems
                item.Paint(g)
            Next item
        End Sub

        'Returns the item whose bounding rectangle contains the specified point.
        'Used for hyperlink hit-testing.
        Public Function LocateItem(ByVal x As Integer, ByVal y As Integer) As HTMLRenderItem
            For Each item As HTMLRenderItem In Me.myRenderItems
                If (x >= item.Left) AndAlso (x <= item.Left + item.Width) _
                    AndAlso (y >= item.Top) AndAlso (y <= item.Top + item.Height) Then Return item
            Next item

            Return Nothing
        End Function

        Public Function LocateItem(ByVal aname As String) As HTMLRenderItem
            For Each item As HTMLRenderItem In myRenderItems
                If item.Element.Name.Length > 0 AndAlso item.Element.Name.CompareTo(aname) = 0 Then Return item
            Next item

            Return Nothing
        End Function

        Public Overloads Overrides Function ToString() As String
            Dim s As String = ""
            s = "<HTMLRenderBand top:" + Me.Top.ToString() + " width:" + Me.Width.ToString() + _
                " height:" + Me.Height.ToString() + " "

            For Each item As HTMLRenderItem In myRenderItems
                s += item.ToString()
            Next item

            s += " HTMLRenderband>"
            Return s
        End Function

    End Class

End Namespace
